modules/library/metadata/src/main/java/org/geotools/factory/CategoryRegistry.java

```diff
@@ -0,0 +1,200 @@
+package org.geotools.factory;
+
+import org.geotools.util.PartiallyOrderedSet;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.geotools.util.Utilities.stream;
+
+/*
+ * TODO: document all
+ *
+ *  - no null checks
+ *  - no subclass relationships between categories and instances
+ */
+
+class CategoryRegistry {
+
+	private final FactoryRegistry factoryRegistry;
+	private final Map<Class<?>, InstanceRegistry<?>> categories = new HashMap<>();
+
+	public CategoryRegistry(FactoryRegistry factoryRegistry) {
+		this.factoryRegistry = factoryRegistry;
+	}
+
+	public void registerCategory(final Class<?> category) {
+		// TODO: should anything happen if the category is already registered?
+		categories.put(category, new InstanceRegistry<>(factoryRegistry, category));
```

Comment:
#### Multiple Registration

I think the original code just `put` the specified provider, i.e. it overwrote the previous one without comment. I'll check which way it was, replicate the behavior and document in JavaDoc.

#### Thread Safety

`FactoryRegistry` Javadoc says:

> NOTE: This class is not thread safe

I think we'd need to do more to guarantee thread-safety than just replacing `Map` with `ConcurrentMap` - I think that should be a separate issue/PR.